tests/suites/test_suite_mpi.function

```diff
@@ -1720,6 +1720,378 @@ exit:
 }
 /* END_CASE */
 
+/* BEGIN_CASE */
+void mbedtls_mpi_core_add_if( char * input_l, char * input_r,
+                              char * input_X4, int carry4,
+                              char * input_X8, int carry8 )
+{
+    mbedtls_mpi X4, X8, l, r;
+    mbedtls_mpi_uint *la = NULL;
+    mbedtls_mpi_uint *ra = NULL;
+    mbedtls_mpi_uint *Xa = NULL;
+    mbedtls_mpi_uint *da = NULL;
+
+    mbedtls_mpi_init( &X4 );
+    mbedtls_mpi_init( &X8 );
+    mbedtls_mpi_init( &l );
+    mbedtls_mpi_init( &r );
+
+    TEST_EQUAL( mbedtls_test_read_mpi( &l, input_l ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &r, input_r ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &X4, input_X4 ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &X8, input_X8 ), 0 );
+
+    /* We only need to work with one of (X4, carry4) or (X8, carry8) depending
+     * on sizeof(mbedtls_mpi_uint)
+     */
+    mbedtls_mpi *X = ( sizeof(mbedtls_mpi_uint) == 4 ) ? &X4 : &X8;
+    mbedtls_mpi_uint carry = ( sizeof(mbedtls_mpi_uint) == 4 ) ? carry4 : carry8;
+
+    /* All of the inputs are +ve (or zero) */
+    TEST_EQUAL( l.s, 1 );
+    TEST_EQUAL( r.s, 1 );
+    TEST_EQUAL( X->s, 1 );
+
+    /* Test cases are such that l <= r, so #limbs should be <= */
+    TEST_ASSERT( l.n <= r.n );
+    TEST_ASSERT( X->n <= r.n );
+
+    /* Now let's get arrays of mbedtls_mpi_uints, rather than MPI structures */
```

Comment:
Why? There's already an array limb in the `mbedtls_mpi` objects. Why not use that? Or, probably better: why use `mbedtls_mpi` objects at all? This function is only testing functions that operate on limb arrays.

Reply:
The MPIs are read very nicely for us by `mbedtls_test_read_mpi()` and yes, give us an array of limbs. However, `mbedtls_test_read_mpi()` doesn't guarantee the size of array we need, and we also want to re-read the data for destructive tests. I find this a convenient and sensible pattern. I'm sorry you disagree.

Reply:
So why use `mbedtls_test_read_mpi` at all, instead of  a function that directly reads limb arrays?

Reply:
Initially, because #6095 (which adds the ability to read limb arrays directly) was happening in parallel. Even with those added, it looks more complicated to do.